src/System.Linq/src/System/Linq/Partition.cs

```diff
@@ -389,5 +389,308 @@ public int GetCount(bool onlyIfCheap)
+                {
+                    using (IEnumerator<TSource> en = _source.GetEnumerator())
+                    {
+                        Debug.Assert(_minIndexInclusive + index >= 0, $"Adding {nameof(index)} caused {nameof(_minIndexInclusive)} to overflow.");
+
+                        if (SkipBefore(_minIndexInclusive + index, en) && en.MoveNext())
+                        {
+                            found = true;
+                            return en.Current;
+                        }
+                    }
+                }
+
+                found = false;
+                return default(TSource);
+            }
+
+            public TSource TryGetFirst(out bool found)
+            {
+                using (IEnumerator<TSource> en = _source.GetEnumerator())
+                {
+                    if (SkipBeforeFirst(en) && en.MoveNext())
+                    {
+                        found = true;
+                        return en.Current;
+                    }
+                }
+
+                found = false;
+                return default(TSource);
+            }
+
+            public TSource TryGetLast(out bool found)
+            {
+                using (IEnumerator<TSource> en = _source.GetEnumerator())
+                {
+                    if (SkipBeforeFirst(en) && en.MoveNext())
+                    {
+                        int remaining = Limit - 1; // Max number of items left, not counting the current element.
+                        int comparand = HasLimit ? 0 : int.MinValue; // If we don't have an upper bound, have the comparison always return true.
+                        TSource result;
+
+                        do
+                        {
+                            remaining--;
+                            result = en.Current;
+                        }
+                        while (remaining >= comparand && en.MoveNext());
+
+                        found = true;
+                        return result;
+                    }
+                }
+
+                found = false;
+                return default(TSource);
+            }
+
+            public TSource[] ToArray()
+            {
+                using (IEnumerator<TSource> en = _source.GetEnumerator())
+                {
+                    if (SkipBeforeFirst(en) && en.MoveNext())
+                    {
+                        int remaining = Limit - 1; // Max number of items left, not counting the current element.
+                        int comparand = HasLimit ? 0 : int.MinValue; // If we don't have an upper bound, have the comparison always return true.
+                        var builder = new LargeArrayBuilder<TSource>(initialize: true);
+
+                        do
+                        {
+                            remaining--;
+                            builder.Add(en.Current);
+                        }
+                        while (remaining >= comparand && en.MoveNext());
```

Comment:
We're dishing out an extra unnecessary write/branch per iteration for the loop if `HasLimit` is false (e.g. if only `Skip` has been called, and not `Take`). It could be specialized like this:

```cs
if (!HasLimit)
{
    do { builder.Add(en.Current); } while (en.MoveNext());
}
else
{
    do { remaining--; builder.Add(en.Current); } while (remaining >= 0 && en.MoveNext());
}
```

I figured it wasn't worth the extra code though.